### PR TITLE
5th assignment, tests are incomplete but in process

### PR DESCRIPTION
**Pull request by kris1319**
I thought about how to generate and to check simply a large random test but didn't find any nice solution.

### CODE REVIEW

File: Assignment_5/alphabet.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+unsigned Alphabet::GetNumberOfLetters() const {
+    return number_of_letters_;
+}
+
+std::vector<char> Alphabet::ToVectorOfLetters() const {
+    return alphabet_;
+}
+
+bool Alphabet::operator==(const Alphabet& other) {
```
Comment (ssancho2) [inline Assignment_5/alphabet.cpp]: Why not `return alphabet_ == other.alphabet_` ? :)

File: Assignment_5/alphabet.cpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            SetEdges(dictionary[i - 1], dictionary[i]);
+        }
+    }
+}
+
+void Alphabet::SetEdges(const std::string& first, const std::string& second) {
+    unsigned skipped = 0;
+    unsigned min_size = std::min(first.length(), second.length());
+    while (skipped < min_size && second[skipped] == first[skipped]) {
+        AddLetterVertex(second[skipped]);
```
Comment (ssancho2) [inline Assignment_5/alphabet.cpp]: I think code will be simpler if you move the AddLetterVertex calls out of this function (and into BuildLetterGraph), either in the current loop or on another new one before it.

File: Assignment_5/alphabet.cpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        skipped++;
+    }
+
+    if (skipped < min_size) {
+        AddLetterVertex(second[skipped]);
+        AddLetterVertex(first[skipped]);
+
+        const auto it = letter_graph_[second[skipped]].find(first[skipped]);
```
Comment (ssancho2) [inline Assignment_5/alphabet.cpp]: Why the check to see if it's there? It's a set, just insert it as many times as you like :)

File: Assignment_5/alphabet_ut.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    Alphabet alph(dict);
+
+    EXPECT_TRUE(alph == Alphabet(alph_val));
+    EXPECT_EQ(alph.GetNumberOfLetters(), alph_val.size());
+}
+
+TEST(Alphabet, OneLetterAlphabetTest) {
+    std::vector<char> alph_val {'a'};
+    std::vector<std::string> dict {
+          "aaaaaa"
```
Comment (ssancho2) [inline Assignment_5/alphabet_ut.cpp]: I don't think this is a valid dictionary, the usual convention on lexicographical order is that if a word is a prefix of another, it comes first, i.e. "a" < "aa", and this is independent of the alphabet/ordering of letters.

**Comment by ssancho2** [comment]
Forgot, about random tests: Yes, indeed it's a bit involved. You can pick the number of letters in the alphabet at random (no need to pick which ones, that's the same), then a random permutation of them. Then generate some random words (shorter words work better I think, and not too many) and sort using the normal alphabet. Then replace all letters using the chosen permutation. 
You also need to copy the code to pick the constrains from the dictionary, so you only test the topological sort part really, but that is the hardest bit.
Then you start wondering if you need to write tests for the test :)

File: Assignment_5/alphabet.cpp (hunk 1 above)
Comment (kris1319) [inline Assignment_5/alphabet.cpp]: Oh, I didn't know that I can compare vectors only with '==' :)

File: Assignment_5/alphabet.cpp (hunk 3 above)
Comment (kris1319) [inline Assignment_5/alphabet.cpp]: Okay :)

File: Assignment_5/alphabet_ut.cpp (hunk 4 above)
Comment (kris1319) [inline Assignment_5/alphabet_ut.cpp]: Yes, I copy-pasted it from the test above and didn't change ordering.

**Comment by kris1319** [comment]
Firstly I'm sorry that I solved a new problem too late. It didn't take so much time I just was out of keyboard :)

Okay, so... The worst case is the case where all N! permutations are valid (where N is the number of letters). And first naive solution is to try all permutations but it takes to much time complexity because for each permutation we should go through it and check order... It is boring and I don't like it.
Another idea is that we can clip some paths in permutations generating that are definitely wrong. So let's "traverse" a letter graph from bottom to up and for each iteration:

1. Choose all sink vertices (vertices with no outgoing edges), assume that now there are N letters/vertices. 

2. For each such node:

    1. Mark current node as used.
    2. Remove all incoming edges to current node.
    3. Call itself recursively with N - 1 letters and get returned list of possible alphabets for subproblem.
    4. Restore edges and mark current node as unused.
    5. If returned list is empty add to the result vector one alphabet consisting only with current letter,
        else append current letter to all returned alphabets and insert modified alphabets into the result vector.

3. Return constructed vector of the possible alphabets.

**Comment by kris1319** [comment]
About generating tests.
Yes, exactly! All my troubles are connected with testing topological sort :) Generate random data is not so hard, but writing complex algorithm for testing TopSort seems to be not so good idea.

What should usually engineers do in the same cases? If the main problem is very important, it needs to cover as much as possible tests and also testing takes a lot of effort do they use mathematical approaches to prove suggested algorithm? 
And I think it also needs to be verified with tests...

**Comment by ssancho2** [comment]
No problem! I really enjoy this :)
Yes, that algorithm makes sense! I'll add a few details:
On ii, you may need to add nodes to the 'sink vertices' , then remove them on iv (Or just recompute this set every time).
It's interesting to think how to store edges to efficiently add/remove them.
The only way for the function to return a empty array is the base case, i.e. when there are 0 nodes, right?

**Comment by ssancho2** [comment]
Regarding testing, sometimes you can split the problem, so here for example you make a function that given the dictionary generates the constrains, and write unit tests for that, then you can reuse that function on a unit test and be confident that it works. Same for the function that checks constrains.
It's also very common to have tests that use a 'golden set' (specially for bigger tests, like integration tests), they just check that the output is the expected value, even if there are other valid answers. This of course may break when the algorithm changes, but it does detect if you break stuff. Just need to be careful to check by hand when you change the 'golden set'.
Sadly mathematical/formal approaches are very rare :(

**Comment by kris1319** [comment]
In my current solution (which have been already pushed) I don't actually remove edges. My idea is that I need only find all sink vertices and they can be found by checking how many outcoming edges they have. So in 'remove incoming edges to the current one node' step I just decrease the count of out coming edges for each parent vertex of the current node. And when the number becomes zero I use the node as a sink node. On the "restoring edges" step I increase the number of out coming edges...
I think it works.

About storing edges.
The first idea that comes to me is to numerate each edge and use extra memory to mark if the particular edge was removed. It can be vector of bools or set of deleted edges. Another way is to store graph not as adjacency list but as list of edges... Also it can be implemented on the current structure with map of sets where any edge can be removed from set and inserted as well.

Hmm... Actually no :( If there are no nodes (letters) it returns empty vector, however if there are nodes BUT the letter graph has cycles like [ v1 -> v2 -> v3 -> v1] there are no sink vertices and method will return empty list too. In more complex cases when we have graph with cycles and also with sink nodes the method will return some invalid or wrong vector.

**Comment by ssancho2** [comment]
Hehe, I haven't noticed you actually wrote the code :) I'll take a look at it.
True, in the case of cycles you will also return the empty array, for mixed cycles/valid graphs, I think your algorithm will just return the valid alphabets ignoring the letters on cycles... and you can detect that just looking at the size of any solution, or if you are about to return the empty vector and N > 0.

**Pull request merged**